jira.go

```diff
@@ -267,9 +289,72 @@ func (c *Client) NewMultiPartRequest(method, urlStr string, buf *bytes.Buffer) (
 	return c.NewMultiPartRequestWithContext(context.Background(), method, urlStr, buf)
 }
 
+func (c *Client) Do(req *http.Request, v interface{}) (*Response, error) {
+	return c.doWithDeadline(req, v)
+}
+
+func (c *Client) doWithDeadline(req *http.Request, v interface{}) (*Response, error) {
+	ctx := req.Context()
+	if failFast(ctx) {
+		return c.do(req, v)
+	}
+
+	deadline, deadlineIsSet := ctx.Deadline()
+	if !deadlineIsSet {
+		deadline = time.Now().Add(retryIntervalDefault)
+		dctx, cancel := context.WithDeadline(ctx, deadline)
+		defer cancel()
+		ctx = dctx
+		req = req.Clone(ctx)
+	}
+	isDeadline := func(t time.Time) bool {
+		return t.Sub(deadline) > 0
+	}
+	errDeadlineWrapped := func(resp *Response, err error) error {
+		return errors.Wrapf(context.DeadlineExceeded, "path: %s, err: %+v, code: %d", req.URL.String(), err, resp.StatusCode)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		resp, err := c.do(req, v)
+		if resp != nil {
```

Comment:
Why not `err != nil`? Are there cases when both are not nil?

Reply:
yes almost everything not 200 has both resp and err not nil.

Reply:
Not if it's calling HTTP directly under the hood:

> A non-2xx response doesn't cause an error. Any returned error will be of type *url.Error. The url.Error value's Timeout method will report true if the request timed out.

Reply:
No it's not called directly, it is post processed and let's keep it like that
https://github.com/lwsanty/go-jira/blob/51ae4740d9270f042221183d1d2171791227fba5/jira.go#L271